streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1486,6 +1726,35 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
   final def transduce[R1 <: R, E1 >: E, A1 >: A, C](sink: ZSink[R1, E1, A1, A1, C]): ZStream[R1, E1, C] =
     transduceManaged[R1, E1, A1, C](ZManaged.succeed(sink))
 
+  /**
+   * Acquire a permit for every consumer task.
+   */
+  final def withPermit(sema: Semaphore): ZStream[R, E, A] =
+    withPermits(sema, 1)
+
+  /**
+   * Acquire n permits for every consumer task.
+   */
+  final def withPermits(sema: Semaphore, n: Long): ZStream[R, E, A] =
+    new ZStream[R, E, A] {
+      def fold[R1 <: R, E1 >: E, A1 >: A, S]: ZStream.Fold[R1, E1, A1, S] =
```

Comment:
This too can be written in terms of `process`

Reply:
Removed as I don't think this make sense with the new process encoding

Reply:
Hey why not? This can still work as:
```scala
ZStream.fromInputStreamManaged {
  self.process.map { as =>
    sema.withPermits(n)(as)
  }
}
```